src/Data/LLVM/BitCode/IR/Function.hs

```diff
@@ -1454,3 +1460,34 @@ getDecodedAtomicRWOp 8  = pure AtomicMin
 getDecodedAtomicRWOp 9  = pure AtomicUMax
 getDecodedAtomicRWOp 10 = pure AtomicUMin
 getDecodedAtomicRWOp v  = Assert.unknownEntity "atomic RWOp" v
+
+{-
+Note [Typing function applications]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The LLVM Language Reference Manual says about the `call` instruction:
+
+    <result> = [tail | musttail | notail ] call [fast-math flags] [cconv] [ret attrs] [addrspace(<num>)]
+               <ty>|<fnty> <fnptrval>(<function args>) [fn attrs] [ operand bundles ]
+
+    ...
+
+    * ‘fnty’: shall be the signature of the function being called. [...]
+    * ‘fnptrval’: An LLVM value containing a pointer to a function to be called. [...]
+
+One slightly surprising aspect of this instruction (at least, this was not
+obvious to me when I first read it) is that `fnptrval` does _not_ have the type
+`fnty`. Rather, `fnptrval` has a pointer type, and the type of the underlying
+memory has type `fnty`.
```

Comment:
> the type of the underlying memory has type `fnty`

nit: I'm not sure how I feel about this statement. Can a chunk of executable code residing in memory be said to have some kind of function type, in the same sense that e.g, a struct stored on the heap has a struct type? Not exactly sure how to improve it either, so feel free to ignore this non-constructive comment :smile:

Reply:
I agree that this is a somewhat flimsy way to describe this. I was deliberately trying to avoid saying "the pointee type is `fnty`", as the phrase "pointee type" is not applicable for opaque pointers. On the other hand, I need a way to describe the property that `fnty` has. There must be _some_ way of doing this... I will see how LLVM's opaque pointers page talks about this.

Reply:
The [opaque pointers page](https://llvm.org/docs/OpaquePointers.html) uses phrasing like this:

> Some instructions still need to know what type to treat the memory pointed to by the pointer as. For example, a load needs to know how many bytes to load from memory and what type to treat the resulting value as. In these cases, instructions themselves contain a type argument.

So LLVM's own phrasing isn't too far off from what I'm using above—they say the type to _treat_ the memory as, rather than the type _of_ the memory. Would you be OK with this framing, @langston-barrett?